WordPress/Classes/Utility/BackgroundTasks/WeeklyRoundupBackgroundTask.swift

```diff
@@ -34,85 +36,142 @@ private class WeeklyRoundupDataProvider {
-                case .failure(let error):
-                    completion(.failure(error))
+    /// Fetches the top site statistics from all available sites in the provided context.
+    ///
+    /// This method retrieves all sites in the given context, then fetches the weekly stats
+    /// for each site. The result, which includes the top 5 sites based on the received stats,
+    /// is returned through the provided completion handler.
+    ///
+    /// **Note:** The completion handler is executed on the context's queue.
+    private func getTopSiteStats(in context: NSManagedObjectContext, completion: @escaping (Result<SiteStats?, Error>) -> Void) {
+        let result: (Result<[Blog], Error>) -> Void = { [weak self] sitesResult in
+            guard let self = self else {
+                return
+            }
+            switch sitesResult {
+            case .success(let sites):
+                guard sites.count > 0 else {
+                    completion(.success(nil))
                     return
+                }
+                self.getTopSiteStats(from: sites, in: context, completion: completion)
+            case .failure(let error):
+                completion(.failure(error))
+                return
+            }
+        }
+        self.getSites(in: context) { sitesResult in
+            context.perform {
+                result(sitesResult)
             }
         }
     }
 
-    private func getTopSiteStats(from sites: [Blog], completion: @escaping (Result<SiteStats?, Error>) -> Void) {
-        var endDateComponents = DateComponents()
-        endDateComponents.weekday = 1
-
-        // The DateComponents timezone is ignored when calling `Calendar.current.nextDate(...)`, so we need to
-        // create a GMT Calendar to perform the date search using it, instead.
-        var gmtCalendar = Calendar(identifier: .gregorian)
-        gmtCalendar.timeZone = TimeZone(secondsFromGMT: 0)!
-
-        guard let periodEndDate = gmtCalendar.nextDate(after: Date(), matching: endDateComponents, matchingPolicy: .nextTime, direction: .backward) else {
-            DDLogError("Something's wrong with the preiod end date selection.")
+    /// Fetches the top site statistics for a given array of blogs.
+    ///
+    /// This method asynchronously fetches weekly statistics for each blog in the provided list.
+    /// After all data has been fetched, it filters out the top 5 sites based on the received stats
+    /// and returns the result through the provided completion handler.
+    ///
+    /// **Note:** The completion handler is executed on the context's queue.
+    private func getTopSiteStats(
+        from sites: [Blog],
+        in context: NSManagedObjectContext,
+        completion: @escaping (Result<SiteStats?, Error>) -> Void
+    ) {
+        guard let periodEndDate = Self.makePeriodEndDate() else {
+            DDLogError("Something's wrong with the period end date selection.")
             return
```

Comment:
❓ Do we need to call the completion handler here?

Reply:
@hassaanelgarem It wasn't in the previous implementation but I agree we should call the completion handler there.

Reply:
Resolved in https://github.com/wordpress-mobile/WordPress-iOS/pull/21076/commits/eebf90cd1ab03d1a2f9f4b38427bdb7a18e03a96